f3/granite.go

```diff
@@ -728,6 +1039,99 @@ func (c *convergeState) findMinTicketProposal() ECChain {
 	return minValue
 }
 
+// BitSet is a bitset implementation.
+type BitSet struct {
+	bits          []uint64
+	size          uint64
+	hammingWeight uint64
+}
+
+func NewBitSet(capacity uint64) *BitSet {
+	return &BitSet{
+		bits:          make([]uint64, (capacity+63)/64),
+		size:          capacity,
+		hammingWeight: 0,
+	}
+}
+
+// Set sets the given bit to 1.
+func (b *BitSet) Set(index uint64) {
+	if b.IsSet(index) {
+		return
+	}
+	segment, bit := index/64, index%64
+	b.bits[segment] |= 1 << bit
+	b.hammingWeight++
+}
+
+// Clear sets the given bit to 0.
+func (b *BitSet) Clear(index uint64) {
+	if !b.IsSet(index) {
+		return
+	}
+	segment, bit := index/64, index%64
+	b.bits[segment] &^= 1 << bit
+	b.hammingWeight--
+}
+
+// IsSet returns whether the given bit is set to 1.
+func (b *BitSet) IsSet(index uint64) bool {
+	segment, bit := index/64, index%64
+	return (b.bits[segment] & (1 << bit)) != 0
+}
+
+// Size returns the number of bits in the BitSet.
+func (b *BitSet) Size() uint64 {
+	return b.size
+}
+
+// HammingWeight returns the number of bits set to 1.
+func (b *BitSet) HammingWeight() uint64 {
+	return b.hammingWeight
+}
+
+// Iterate calls the given function for each set bit in the BitSet.
+func (b *BitSet) Iterate(callback func(index uint64)) {
+	for i := uint64(0); i < b.size; i++ {
+		if b.IsSet(i) {
+			callback(i)
+		}
+	}
+}
+
+// SenderIndex maps ActorID to a unique index in the range [0, len(powerTable.Entries)).
+// This is used to index into a BitSet. index2Actor is the reverse mapping.
+type SenderIndex struct {
+	Actor2Index map[ActorID]uint64
+	index2Actor []ActorID
+}
+
+func NewSenderIndex(table PowerTable) *SenderIndex {
+	senderIndex := &SenderIndex{
+		Actor2Index: make(map[ActorID]uint64, len(table.Entries)),
+		index2Actor: make([]ActorID, len(table.Entries)),
+	}
+	keys := make([]ActorID, 0, len(table.Entries))
+
+	// Extract keys from the map
+	for senderID := range table.Entries {
+		keys = append(keys, senderID)
+	}
+
+	// Sort the keys
+	sort.Slice(keys, func(i, j int) bool { return keys[i] < keys[j] })
```

Comment:
This probably isn't the ordering we will want - probably want descending by power instead.

Reply:
done.